Added .serialized export

tufPenny opened the pull request:
Serialized took some time because of the sparse information about it out there but it now should work (with all limitations of Mitsuba and Assimp: Only 2D texture coordinates, only one set of them per mesh, only single precision indices, only rgb vertex colors). In my tests it was faster with a much smaller file size.
I'd love for someone else to check if zlib really gets imported correctly on every os. There is a bit of a chaos in the commits when I tried to get it to work...

tufPenny committed:
added bump map

Mitsuba can load both a height/bump map AND a normal map, therefore I implemented this loader to be able to load both, should they both exist.

tufPenny committed:
cleanup

added `xmlBrdfMapWrapper` to keep `toXML` clear and easy to read.

tufPenny committed:
Merge branch 'hesom:master' into master

tufPenny committed:
Added zlib

The compression in the mitsuba 3 serialized format requires this.

tufPenny committed:
Revert "Merge branch 'hesom:master' into master"

This reverts commit 6d075d0f83f0e8da4a1db0de7f4272404d0687c2, reversing
changes made to 72440b3f3e781df2df9791696d72f0fffeb2d1b9.

tufPenny committed:
corrected submodule zlib

tufPenny committed:
Revert "corrected submodule zlib"

This reverts commit 869f5aa91245613195cab010cd08d7836051edae.

tufPenny committed:
fixed fmt

god, too many dumb commits of mine created quite the mess

tufPenny committed:
started work on serialized export

tufPenny committed:
first version of serialized exporter done

this is completely untested, but it does compile

tufPenny committed:
faster file writing

after reading the docs of streams this should write pretty much the entire file in one go
plus I added some TODOs

tufPenny committed:
prevented reallocations of enflatedData

for large models probably a significant speedup

tufPenny committed:
corrected metadata writing into the file

plus some debugging, this is still not working right

tufPenny committed:
fixed the compression

Mitsuba can now read the file without throwing an error, however the result doesn't get rendered correctly so something is still wrong.

tufPenny committed:
exporting .serialized meshes works now

some cleanup still neccessary

tufPenny committed:
Added duplicate face removal for serialized

additional cleanup

tufPenny committed:
additional cleanup

removed some debugging artifacts

tufPenny committed:
even more cleanup

hesom commented on kontsuba/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 add_library(kontsuba_core STATIC
     core/converter.cpp
 )
+find_package(ZLIB REQUIRED)
```
assimp already builds zlib, so it should be just available. find_package only finds zlib in the system, but I would prefer to use the zlib build by assimp. Can you try to remove this line in see if it stills works?

tufPenny replied:
When removing this line it sadly does not work.

hesom replied:
And with "zliblibraries" or "${ZLIB_LIBRARIES}" instead of "ZLIB::ZLIB"?

tufPenny replied:
Interestingly removing "find_package(ZLIB REQUIRED)" and modifying "PRIVATE ZLIB::ZLIB" to "PRIVATE ${ZLIB_LIBRARIES}" works under my wsl ubuntu system via cmake, but "pip install ." does not work anymore.

hesom commented on kontsuba/CMakeLists.txt:
```diff
@@ -12,6 +13,7 @@ target_link_libraries(kontsuba_core
     PRIVATE tinyxml2
     PRIVATE tinyply
     PRIVATE fmt
+    PRIVATE ZLIB::ZLIB
```
This should stay (might be ZLIB::zlib)

tufPenny committed:
Added command line arguments

Used for switching handedness or UV origin
Started work on better materials

tufPenny committed:
added forceTwoSided argument

an imported model will therefore be twosided if assimp sets the flag or if the user specifies in command line arguments
improved the flags with a much more readable (and less prone to errors) enum

tufPenny committed:
first try to improve material import

still buggy, but a first step

tufPenny committed:
incremental improvements

tufPenny committed:
a little cleanup
